### ISSUE TITLE
Psionic Runtime: codify the Compiled, Allocator, Program, and Compiler split

### ISSUE THREAD

**Issue by AtlantisPleb** [OpenAgentsInc/openagents#3604]
## Why Now

Tinygrad's runtime is unusually legible because the core runtime split is
explicit:

- `Compiled`
- `Allocator`
- `Program`
- `Compiler`

The Tinygrad parity audit concludes that Psionic should match that level of
runtime clarity, not because the names are sacred, but because the boundaries
are what make backend work, memory ownership, program loading, and execution
truth inspectable.

Psionic docs already describe runtime and backend crates, but the backlog still
lacks one issue that codifies this split as a framework-completion target.

## Current Baseline

- `crates/psionic/docs/ARCHITECTURE.md` already names runtime and backend
  responsibilities at a high level.
- Existing inference issues focus on server, scheduler, KV, and routing layers.
- There is not yet a dedicated issue to lock down the reusable runtime contract
  that backend crates are expected to satisfy.

## Scope

Codify and land the core Psionic runtime contract.

This issue should cover:

- the stable split between device bootstrap, allocation, compiled program
  ownership, and compiler responsibilities
- canonical device-opening and device-canonicalization rules
- buffer lifecycle and allocation semantics, including reusable allocator posture
- explicit policy for zero-copy or external-memory interop boundaries where
  supported
- program load and invocation contracts that later compiler and scheduler issues
  can target without backend-specific special cases
- at least one caching allocator or LRU-style reuse posture where that is the
  right substrate choice

## Deliverables

- Runtime traits or equivalent contracts for the split above.
- At least one reference backend and one non-reference backend implementing the
  contract honestly.
- Tests for buffer lifecycle, allocation reuse, and program invocation.
- Doc updates if the runtime contract changes the current Psionic architecture.

## Acceptance Criteria

- Backend crates can implement one reusable runtime contract instead of each
  inventing its own execution vocabulary.
- Allocation, program loading, and execution boundaries are explicit and typed.
- Buffer lifetime and reuse behavior are inspectable and testable.
- The runtime split is clear enough that later scheduler, compiler, and
  multi-device issues can depend on it without redefining the same abstractions.

## Dependencies

- none

## Design Constraints

- Keep ownership in reusable Psionic layers; do not push runtime truth into app
  crates.
- Do not make risky zero-copy or raw-pointer interop the default product path.
- Preserve room for proof-bearing execution and capability truth.

## References

- `docs/audits/2026-03-14-tinygrad-parity-target-for-psionic-audit.md`
- `crates/psionic/docs/ARCHITECTURE.md`

**Comment by AtlantisPleb** [comment]
Implemented on \ through the now-explicit runtime/compiler split.

Landed surfaces:
- \: explicit \, \, and \ traits plus runtime resource/cache reporting
- \: \ and compiler pipeline entrypoints over Psionic IR

Validation run on current \:
- \
running 95 tests
test parity::tests::default_policy_keeps_cpu_as_reference_and_quantized_budget_looser ... ok
test parity::tests::generation_budget_is_exact_when_seeded_and_semantic_when_unseeded ... ok
test parity::tests::logits_comparison_allows_one_rank_drift_for_quantized_paths ... ok
test parity::tests::embedding_comparison_uses_budget_and_cosine_similarity ... ok
test gpt_oss::tests::gpt_oss_decode_graph_matches_llama_cpp_high_level_order ... ok
test gpt_oss::tests::decode_graph_signature_changes_with_layer_count ... ok
test proof::tests::sandbox_bundle_carries_sandbox_context_and_artifact_digests ... ok
test proof::tests::activation_fingerprint_posture_and_ref_are_explicit ... ok
test activation_fingerprint::tests::quantized_adapter_verifies_tolerant_bucket_matches ... ok
test activation_fingerprint::tests::quantized_adapter_emits_self_consistent_artifact ... ok
test activation_fingerprint::tests::quantized_adapter_benchmark_reports_iterations_and_costs ... ok
test structured_output::tests::gbnf_matcher_reports_open_ended_grammar_can_continue ... ok
test structured_output::tests::choice_matcher_materializes_discrete_values ... ok
test structured_output::tests::gbnf_matcher_accepts_literal_choices_and_prefixes ... ok
test structured_output::tests::json_object_matcher_accepts_nested_json ... ok
test structured_output::tests::local_structured_output_capability_inventory_is_stable ... ok
test structured_output::tests::local_structured_output_parser_inventory_is_stable ... ok
test structured_output::tests::json_schema_lowerer_supports_small_objects ... ok
test proof::tests::cluster_execution_changes_bundle_digest ... ok
test proof::tests::runtime_manifest_distinguishes_identity_from_mutable_variables ... ok
test tests::accelerator_requirements_can_match_exact_delivery ... ok
test tests::accelerator_requirements_can_detect_underdelivery ... ok
test structured_output::tests::tagged_structure_matcher_materializes_tag_and_value ... ok
test tests::accelerator_requirements_mark_topology_change_as_compatible_substitution ... ok
test tests::backend_health_tracker_records_only_real_changes ... ok
test structured_output::tests::regex_matcher_supports_prefix_and_materialization ... ok
test tests::backend_extension_support_serializes_stably ... ok
test tests::amd_backend_model_serializes_mode_topology_risk_and_recovery ... ok
test tests::backend_selection_can_carry_cuda_backend_identity ... ok
test tests::backend_runtime_resources_serialize_stably ... ok
test tests::backend_selection_can_surface_multi_device_topology_truth ... ok
test proof::tests::proof_bundle_digest_is_stable_for_identical_inputs ... ok
test tests::backend_selection_can_surface_refusal_retry_and_same_backend_slow_path ... ok
test tests::backend_selection_selected_device_inventory_uses_runtime_budget_bytes ... ok
test tests::buffer_handles_can_distinguish_quantized_storage_from_dequantized_fallback ... ok
test tests::cache_observation_serializes_stably ... ok
test tests::cluster_compute_market_trust_assessment_digest_changes_with_refusal_shape ... ok
test tests::backend_selection_helpers_capture_direct_and_fallback_truth ... ok
test tests::cluster_compute_market_trust_assessment_round_trips_with_stable_digest ... ok
test tests::cluster_execution_capability_profile_digest_changes_when_cache_truth_changes ... ok
test tests::cluster_execution_capability_profile_digest_changes_when_lane_support_changes ... ok
test tests::cluster_execution_capability_profile_digest_changes_when_residency_truth_changes ... ok
test tests::cluster_execution_capability_profile_digest_changes_when_serving_semantics_change ... ok
test tests::cluster_execution_capability_profile_round_trips_with_stable_digest ... ok
test tests::communication_eligibility_can_be_derived_from_capability_profile ... ok
test tests::delivered_execution_context_can_carry_training_collective_context ... ok
test tests::backend_selection_can_publish_cluster_compute_market_trust_assessment_truth ... ok
test tests::backend_selection_can_publish_declared_cluster_capability_profile_truth ... ok
test tests::delivered_execution_context_can_carry_cluster_evidence ... ok
test tests::delivered_execution_context_prefers_replicated_cluster_topology ... ok
test tests::delivered_execution_context_can_carry_training_recovery_context ... ok
test tests::delivered_execution_context_surfaces_layer_sharded_handoffs ... ok
test tests::delivered_execution_context_surfaces_tensor_sharded_collectives ... ok
test tests::device_inventory_qualifiers_mark_mig_devices_as_partitioned ... ok
test tests::execution_capability_profiles_are_machine_checkable ... ok
test tests::fallback_lattice_supports_refuse_replan_retry_and_same_backend_slow_path ... ok
test tests::generation_scheduler_metrics_track_fallback_counts ... ok
test tests::kv_cache_state_and_growth_serialize_stably ... ok
test tests::kv_page_layout_reports_page_and_byte_geometry ... ok
test tests::kv_residency_accounting_serializes_stably ... ok
test tests::lane_communication_eligibility_refuses_undeclared_lane_even_when_profile_exists ... ok
test tests::loaded_model_residency_tracks_keepalive_and_request_activity ... ok
test tests::local_serving_isolation_policy_helpers_are_stable ... ok
test tests::mock_runtime_reports_device_and_executes_plan ... ok
test tests::model_admission_can_evict_oldest_idle_model_to_fit_budget ... ok
test tests::local_runtime_observability_serializes_stably ... ok
test tests::model_admission_refuses_when_only_active_models_block_the_budget ... ok
test tests::nvidia_backend_model_serializes_topology_risk_and_recovery ... ok
test tests::paged_tensor_storage_plan_serializes_byte_window_and_page_counts ... ok
test tests::prefix_cache_identity_and_policy_serialize_stably ... ok
test tests::quantization_dispatch_prefers_grouped_block_for_grouped_ggml_workload ... ok
test tests::quantization_support_surfaces_storage_path_and_pending_execution_truth ... ok
test tests::quantized_decode_policy_keeps_latency_sensitive_work_unbatched ... ok
test tests::residency_policy_and_memory_plan_serialize_stably ... ok
test tests::runtime_dispatch_plan_batches_data_plane_work_and_reduces_cost ... ok
test tests::runtime_model_storage_truth_distinguishes_paged_blobs_from_copies ... ok
test tests::runtime_transition_log_retains_latest_events ... ok
test tests::sampling_penalties_honor_the_bounded_lookback_window ... ok
test tests::sampling_policy_serializes_supported_generation_controls ... ok
test tests::sandbox_execution_capability_profiles_are_machine_checkable ... ok
test tests::sandbox_execution_evidence_serializes_stably ... ok
test tests::sandbox_execution_profile_digest_changes_with_bounds ... ok
test tests::seeded_token_sampler_replays_draws ... ok
test tests::served_artifact_identity_digest_changes_when_identity_changes ... ok
test tests::sharded_model_manifest_refuses_partition_mismatch ... ok
test tests::sharded_model_manifest_validates_replicated_layer_and_tensor_topologies ... ok
test validation::tests::backend_probe_references_expose_current_amd_claims ... ok
test tests::signed_cluster_evidence_bundle_refuses_tampered_payload ... ok
test validation::tests::metal_gpt_oss_fallback_maps_to_explicit_refusal_claim ... ok
test validation::tests::metal_gpt_oss_text_generation_maps_to_explicit_matrix_claim ... ok
test validation::tests::minimum_hardware_validation_matrix_covers_required_lanes ... ok
test validation::tests::non_gpt_oss_metal_text_generation_keeps_dense_claim ... ok
test validation::tests::non_matrix_backend_product_pairs_are_marked_not_yet_validated ... ok
test validation::tests::shipped_served_product_claims_map_to_the_minimum_matrix ... ok
test tests::signed_cluster_evidence_bundle_round_trips_and_verifies ... ok

test result: ok. 95 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
- \
running 5 tests
test tests::compile_graph_for_selection_uses_explicit_execution_topology ... ok
test tests::compile_graph_preserves_backend_extension_payloads ... ok
test tests::compile_graph_lists_expected_steps ... ok
test tests::compile_graph_preserves_deterministic_digest ... ok
test tests::compile_graph_with_topology_changes_digest_when_sharding_changes ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

Closing as implemented.
